Fix Chrome number input backspace and invalid input issue

nhunzaker opened the pull request:
In Google Chrome, backspacing an controlled and uncontrolled number inputs causes an unexpected behavior:

![backspace](https://cloud.githubusercontent.com/assets/590904/17160637/6dfbca3a-5374-11e6-9a0d-0a2c52813eea.gif)

More cases (and what fixes look like) here: https://github.com/facebook/react/issues/7253#issuecomment-236074326
## Issues with uncontrolled inputs

This is because assigning `defaultValue` causes side-effects on `value`. It seems to trigger validation on `value` as if it were directly assigned itself. 

This Codepen provides a step by step reproduction case:

http://codepen.io/nhunzaker/pen/zBaokp

![wat](https://cloud.githubusercontent.com/assets/590904/17160341/4d3acf64-5372-11e6-9871-64c6063b23b0.gif)

This PR wraps the `defaultValue` assignment in a conditional so that it will only ever be assigned if the defaultValue has changed.
## Issues with controlled inputs

It looks like the issue here is that value is being assigned using the standard React method of updating attributes when the props get passed in here:

https://github.com/facebook/react/blob/master/src/renderers/dom/shared/ReactDOMComponent.js#L926

There is some logic in ReactDOMInput to prevent duplicate values from being assigned here:

https://github.com/facebook/react/blob/master/src/renderers/dom/client/wrappers/ReactDOMInput.js#L201-L210

Unfortunately the new value has already been assigned.

This PR adds a check for the `value` prop, and only assigns it if is different. It never assigns the value attribute if it hasn't changed:

I don't really know why this works. I didn't think node.setAttribute('value', value) was meaningful. What do you think?

---

Related to https://github.com/facebook/react/issues/7253

zpao commented on src/renderers/dom/shared/DOMPropertyOperations.js:
```diff
@@ -158,6 +158,10 @@ var DOMPropertyOperations = {
+        } else if (attributeName === 'value') {
+          if (node.value != value) {
+            node.value = '' + value;
+          }
```
Ironically our code used to look a lot like this… https://github.com/facebook/react/blob/0.14-stable/src/renderers/dom/shared/DOMPropertyOperations.js#L187-L197

@jimfb - I think this is more fallout from your changes. Can you take a look and figure out a way forward?

nhunzaker replied:
Ack! I crossed the streams!

This PR was supposed to be solely for fixing `defaultValue`, but it looks I got my branches mixed up and pushed the more general fix for both inputs. I actually just replied here:

https://github.com/facebook/react/issues/7253#issuecomment-235602919

Again, pushing to this branch was a total accident. A bunch of tests fail now. I'll work on getting those to pass in the mean time, but could definitely use some advice.

nhunzaker commented:
@zpao Thanks for the historical insight. All tests pass. I believe I've caught every edge case. I'll enumerate them in #7253. I'm happy to add back the `HAS_SIDE_EFFECTS` constant, however this seems to be truly specific to the `value` attribute. What do you think?

nhunzaker commented:
Cases enumerated here:

https://github.com/facebook/react/issues/7253#issuecomment-236074326

nhunzaker commented:
Ah and it looks like `lint` is going to fail because my solution relies on type coercion. Hmm. Do I have permission to ignore these rules?

nhunzaker commented on src/renderers/dom/shared/DOMPropertyOperations.js:
```diff
@@ -158,6 +158,10 @@ var DOMPropertyOperations = {
         } else if (propertyInfo.hasBooleanValue ||
                    (propertyInfo.hasOverloadedBooleanValue && value === true)) {
           node.setAttribute(attributeName, '');
+        } else if (attributeName === 'value' && node.hasAttribute('value')) {
```
Type coercion solved my problems, however it makes the initial case hard. Empty inputs are always equal to `0`. 

I think this could be better. Any suggestions?

nhunzaker commented:
Added ignore statements for ESLint rules and a test for `setAttributeForProperty`.

nhunzaker commented on src/renderers/dom/shared/DOMPropertyOperations.js:
```diff
@@ -158,6 +158,12 @@ var DOMPropertyOperations = {
+          // Use loose coercion to prevent replacement on comparisons like
+          // '3e1' == 30 in Chrome (~52).
+          if (node.value != value) { // eslint-disable-line
+            node.setAttribute(attributeName, '' + value);
```
One thing I'm not clear on yet: _Why assign the `value` attribute at all?_. 

`ReactDOMInput` will assign the `value` property in `updateWrapper`. We could do that in `mountWrapper` as well. Is it important that the attribute be assigned?

Aweary replied:
So, there's a difference between setting `node.value` and calling `node.setAttribute('value')`.  `node.value` tracks the value of the node in JS, and `setAttribute` will actually set the attribute on the element in the DOM.

This change causes an issue where the DOM attribute is never updated, because `node.value` is already updated prior to this check, so it never actually sets the attribute.

nhunzaker replied:
Hmm. So actually: _should the value update for anything other than `checkbox` and `radio`?_

Following browser behavior, editing the text inside of a text input doesn't update the `value` attribute. 

Possible my current codes doesn't do this, but to me that's the correct functionality.

Aweary replied:
React does deviate from browser behavior when it comes to controlled components (e.g., change events on keydown instead of on blur). Currently the `value` attribute does update on controlled inputs, and it looks like we're explicitly setting it, so there may be a good reason I'm not aware of.

cc @spicyj @zpao

nhunzaker replied:
There is an extra complication here... Assigning the `value` attribute forces validation for number inputs. Check this out:

![control](https://cloud.githubusercontent.com/assets/590904/18232983/71c54192-72a9-11e6-8f1f-6141e68d8746.gif)

Notice how the decimal gets stripped? My fix focuses on avoiding touching the input as much as possible to avoid unexpected behavior like this. Fortunately, I believe this also conforms to the standard way inputs work.

spicyj replied:
I believe the current behavior of setting the attribute is intentional though I could be misremembering. I know we definitely need to set `value` on the `<progress>` element. You can look at the history of ReactDOMInput to see when jimfb last made changes.

nhunzaker replied:
Looks like this is taken care of, because `node.value` on progress is never assigned, so it's out of sync with `props.value`, and it reassigns the attribute.

![progress](https://cloud.githubusercontent.com/assets/590904/18248537/cc559d62-7346-11e6-9699-6d7b8b9e1913.gif)

STRML commented:
Anything this is waiting on? This is a pretty obnoxious bug. Thanks so much to @nhunzaker for putting this PR together.

Aweary commented:
I'll try and get this reviewed when I get back from vacation next week, if no one else gets a chance to do it in the meantime. Overall it looks good to me though!

nhunzaker commented:
@Aweary No worries -- I'm on vacation too! Looking forward to getting this one in...

Aweary commented:
@nhunzaker thanks for doing that. Today is my first day back from vacation so I'll try to review it soon. I just want to pull this down and step through the code path with a few examples to make sure this makes sense.

This seems to be an area where we've had a lot of issues in the past, so I just want to be extra careful in avoiding any regressions.

nhunzaker commented:
@Aweary Yep, I totally understand. This is a complicated issue. 

It would be cool if we could build up a test suite for browser quirks. I started working on one here as I worked through these input issues:

https://github.com/nhunzaker/react-issue-reproducer

Hardly satisfactory, and I'm not suggesting you all do anything with it, but I'd love to see browser coverage for inputs, particularly for the change event behavior and side-effect property assignments. So many input issues have test cases written up as jsfiddles, it seems like you could collect them and write webdriver scripts to reproduce the issue.

Aweary commented:
Had some time to review, good work overall, but this introduces an issue where the `value` attribute is never updated in the DOM (see https://github.com/facebook/react/pull/7359/files#r77446117). I'm pretty sure we still want the `value` attribute to stay in sync with `node.value` so we'll have to address that somehow.

nhunzaker commented:
Pushed an update.

I didn't like how there was an exception just for `value` in `setValueForProperty`, so I moved it into a mutation method. I also added a few more tests and caught an edge cause where an empty value switches to zero and the attribute isn't updated (because of the loose check).

Pushing the commit also murdered our conversation thread... So just to clarify:

**All non-input elements always get a fresh `value` attribute when the prop changes**

For example, a `<progress>` element will always get an attribute assignment

```html
<progress value={value} max="1" />
```

![progress](https://cloud.githubusercontent.com/assets/590904/18250594/5002d520-7352-11e6-8b6a-0b1a46538334.gif)

**All non-text input elements always get a fresh `value` attribute when the prop changes.**

Like changing a `radio`'s value will always reflect in the DOM:

```html
<input type="radio" value={true}  />
```

![radio](https://cloud.githubusercontent.com/assets/590904/18250621/92d66cfe-7352-11e6-9aad-db17265d3488.gif)


**Text input elements only assign the `value` attribute once, when the input is created.**

```html
<input type="number" value={value}  />
```

They also only assign a new `value` property if it is loosely different. This prevents manipulation of the input that would expand `3e1` into `30`, or have `3.` drop the decimal place and set the input value to `3`. This makes `number` inputs useable:

![number](https://cloud.githubusercontent.com/assets/590904/18250638/abbcd3d4-7352-11e6-8467-2caca6997cd7.gif)

Aweary commented:
cc @jimfb, are there good reasons for updating the `value` attribute on text inputs for controlled components?

jimfb commented:
@Aweary I think the answer is "YES".  I know most of the cases are handled by properly setting the `defaultValue` property (like the reset button - I think responds to `defaultValue`), but a few things respond to the attribute (off the top of my head, I don't remember which).  Also need to consider compatibility with various browser extensions that read/write the input fields (like password managers, but even the chrome devtools fall into this category because they only show the attributes).  I don't remember exactly, you'd have to test all the input behaviors in all the browsers, and decide which functionality we're willing to sacrifice, but I vaguely remember us running into situations for which setting the attribute was desirable.  Bottom line: Unless there are visible negative effects to setting the attribute, we should probably continue to set it, because you'll get better compatibility that way.

nhunzaker commented:
Actually, unless I've missed something, assigning the `value` attribute sets `defaultValue`:

```javascript
var input = document.createElement('input')
input.defaultValue = '10'
input.value = 5
input.setAttribute('value', 20)
console.log(input.defaultValue) // 20
```

The reset button is actually broken (in Chrome, anyway) right now because assigning value as an attribute changes `defaultValue`, so it always assigns a new reset state:

![reset](https://cloud.githubusercontent.com/assets/590904/18252681/a4ead9b0-735d-11e6-9bfd-fdee888cda8e.gif)

Check it out here:
http://natehunzaker.com/react-issue-reproducer/cases/form-reset/

Code viewable here:
https://github.com/nhunzaker/react-issue-reproducer/blob/gh-pages/cases/form-reset/index.html

---

There's also the additional concern with number inputs: Whenever a number input's value attribute is set to an invalid state, and the user has given input, it raises a validation error and reformats the input's value:

![validated](https://cloud.githubusercontent.com/assets/590904/18252803/838e0dea-735e-11e6-89d8-068c87d79798.gif)

Additionally, assuming the Chrome inspector is representing the DOM correctly, text manipulation by the user does not have an impact on the value attribute.

Aweary commented:
> There's also the additional concern with number inputs: Whenever a number input's value attribute is set to an invalid state, and the user has given input, it raises a validation error and reformats the input's value:

Is there a way we can only call `setAttribute` when the `value` is valid?

nhunzaker commented:
@Aweary Yes, however it is not supported in IE9 (https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement)

nhunzaker commented:
@Aweary I'm sorry. I meant to refer to `input.checkValidity()`

nhunzaker commented:
Also. It's Labor Day! I don't want to take you two away from your holiday weeks any more than I already have. I'm happy to continue this discussion tomorrow.

jimfb commented:
> The reset button is actually broken (in Chrome, anyway) right now because assigning value as an attribute changes defaultValue, so it always assigns a new reset state

@nhunzaker That is not broken behavior - that is the correct behavior.

A reset button will reset the internal value state of the button.  A controlled input has no internal state, and thus the reset button does nothing.  Another way of looking at this: When the controlled input tag rerenders, the result of a rerender should be identical to the result of the input tag when initially rendered with the same data.  This includes the behavior of the reset button (which does nothing if you clicked the button immediately after refreshing the page).

When using a controlled input, an application needs to explicitly listen to the reset button if it wants to perform some action (like clearing one or more inputs) based on that action.  This is expected behavior for controlled inputs, who's value is never supposed to deviate from the controlled value that was passed in.

Once upon a time, we had a bug where the reset button would sometimes clear the value of a controlled input.  That was the bug.  The fact that the value attribute also sets defaultValue is both expected and correct behavior.

nhunzaker commented:
Got it. I was also mistaken that resetting a form would trigger a change event on the input and go through the standard controlled input flow.

I can see now that the input is in a state not represented by the component, which totally makes sense. Though I'm sort of curious what it would look like to (purely day-dreaming) assign a `defaultValue` prop to a controlled input that would control the attribute. The input would revert to that state when the form reset. In that way, controlled inputs would be an enhancement on top of regular uncontrolled inputs, instead of being separate concepts. 

But day dreaming aside... **what do we do about this attribute issue on number inputs**? 

The trouble is that assigning an invalid attribute to a number field causes Chrome to chop off decimal places on numbers and raises a bunch of validation warnings as invalid attributes are assigned to an input's value attribute. This is one of the issues I am trying to solve.

Does Chrome consider this a bug? Should the docs state that validated input fields are incompatible with controlled number inputs (or at least have these known issues)?

Or is the answer to add an extra layer to check for bad inputs, and only assign the `value` attribute  if the field is valid, as @Aweary mentioned? 

What sort of problems could we anticipate with that approach?

jimfb commented:
> Or is the answer to add an extra layer to check for bad inputs, and only assign the value attribute if the field is valid, as @Aweary mentioned?

I think this is the correct answer.

 > What sort of problems could we anticipate with that approach?

Haha, probably lots of problems :P.  But not because the approach is wrong.  Mostly just because the DOM is terrible, and DOM inputs are among the hairiest pieces of the DOM.

nhunzaker commented:
Oh boy... here goes..

```javascript
if (!node.validity || node.validity.badInput === false) {
  node.setAttribute('value', '' + value);
}
```

Concise, but it looks like `3.` is valid input, where as something like `3e` is marked as "bad input".

![Ugh](https://media.giphy.com/media/26uf89X5hSWNLyJgI/giphy.gif).

Alright, time to see if we can consistently, programmatically, get the regexes for the input validator.

nhunzaker commented:
Ack. Rebased this with master and it removed the "needs-revision" flag. This definitely needs a revision!

nhunzaker commented:
Assigning a value attribute to a number input in Chrome is... _hard_. Barring some some change to Chrome, I don't know how to move forward with assigning an attribute/defaultValue on every change. Number inputs simply don't want to behave.

Just to further document the complexity here. Regarding only assigning an attribute on validity:

1. Chrome considers `'3.'` valid input when you type it. However assigning `'3.'` raises a warning and clears the input
2. Interestingly, setting `'3.'` as an attribute _does_ work, however it causes the input's text value to drop the decimal place. Not useful either....
2. A number input of `<input value="3.14" step="0.01" />` still drops the decimal point when backspacing because it considers it `3.` valid, assigning a  new `value` attribute. But again, when the attribute gets assigned, the number input does some magic that kills the decimal place.

Basically, it's both extremely hard to determine if invalid input is handleable, and valid input still leaves the possibility of unwanted behavior. Additionally, the key validator that I want to use, `badInput`, considers trailing decimal places okay (though apparently `input.value = '3.1'`  doesn't).

I'm also afraid that, accounting for Chrome's validity behavior, we might introduce other browser bugs with the extra complexity. This is pure speculation, but I haven't even factored in _custom_ validation behavior like `pattern`. 

So maybe back to the drawing board? Should we file bugs with Chrome, or get any advice from their team? Right now it seems like number inputs are simply uncontrollable, and setting an attribute on change for a validated field will always raise warnings in the console.

Aweary commented:
> So maybe back to the drawing board? Should we file bugs (are these bugs?) with Chrome, or get any advice from their team? 

Have you tested any other browsers with this? If so, do they exhibit the same behaviors?

nhunzaker commented:
Here's what I'm seeing using a basic pen:

http://s.codepen.io/nhunzaker/debug/zKGVEA

| Browser        | `'3.'` as property | `'3.'` as attribute           | `console.log(input.value)` | 
| -------------- | ---------------- | ----------------------------- | -------------------- |
| Chrome 52      | Clear input text, warn | Set attribute to '3.', drop display value decimal place, warn invalid | `"3"` |
| Firefox Dev 50 | Clear input text, no warning | Set attribute, display value remains the same, no warning | `"3"` |
| Safari 9.1.2   | Sets value and display text to '3.'  | Sets '3.' as value attribute, no warn or display value update | `"3."` |
| IE 11          | Sets value and display text to '3.' | Sets '3.' as value attribute, no warn or display value update | `"3."` |
| MS Edge        | Sets value and display text to '3.' | Sets '3.' as value attribute, no warn or display value update. | `"3."` |

So Firefox and Chrome are sort of similar, Safari and IE are the same. Chrome is the only browser that modifies the display value based on attribute assignment.

nhunzaker commented:
No work arounds (or even Chrome specific hacks) discovered yet. I have two ideas, but wanted to run them by you because the labor involved is not trivial. I could also use advice and if the approach is even something the team would consider:

**Idea 1: Update the `value` attribute when the user blurs the input**

React could listen to `blur`, then conditionally set the attribute if the input is valid.

This at least postpones the issue of dropping decimal places when a number loses its decimal places until the user is done entering content.

**Idea 2: Don't assign a `value`attribute for inputs that manipulate `input.value`. Listen to `form.reset()` and other possible mutators of form inputs**

I don't believe `form.reset()` invokes a change event on inputs, but it definitely triggers a `reset` event. React could listen to the form and override the value assignment similarly to how it already overrides inputs without a change handler (like `<input value="static-value" />`

Beyond `form.reset()`, @jimfb hinted at a few browser extensions that might be problematic? Do we know what those are? I'm not interested specifically in making exceptions for them, but maybe they can help me better understand the problem.

nhunzaker commented:
In the mean time, would it be okay to send out a PR that only ships the fix for uncontrolled inputs? I think it just needs the statement starting here:

https://github.com/facebook/react/pull/7359/files#diff-1b0e3b280991cd0b7cde194b30353b3eR218

yaycmyk commented:
I'm still not clear on why we'd want to ever change `defaultValue` after mount in the first place. If the user clears out the input, it would be blank... so there's no point in setting that DOM property more than once.

nhunzaker commented:
> I'm still not clear on why we'd want to ever change defaultValue after mount in the first place.

@yaycmyk Yes. I agree completely, but it's a hard. 😄.

This _is_ ideal for my PR. Touching value stuff on number inputs in Chrome causes them to validate, dropping decimal places and making the cursor jump to the beginning of the input (because number inputs also don't support the selection API 😢)

I need help. I see where I want to go, but I haven't figured out how to get there. Here are the issues I've run into:

1. `form.reset()` breaks controlled inputs because it reverts inputs to their defaultValue but does not alert inputs that they have changed (please correct me if I'm wrong here). Presently, assigning `defaultValue` on every change works around that.
2. `defaultValue` or the `value` attribute are utilized by browser extensions. This has been raised in this thread, though no specific extensions have been identified. I haven't had the time to track these down, I would like to know more...

So if we were to go down this route, **we need to handle `form.reset()`, and figure out specific problems introduced by browser extensions**. Ideally we document them too, for future bug fixers. Otherwise controlled inputs get out of sync and React never gets alerted.  The challenge here is that inputs seem to change their value _after_ the `reset` event propagates. You can't hijack `form.reset()`. I'm still figuring that part out...

Personally, I'd like to see support for `form.reset()` on controlled inputs. Maybe allow controlled inputs to configure their `defaultValue` manually. That would be super cool, and also address number input issues because `defaultValue` wouldn't get assigned again as of #7750. Presently, controlled inputs are a bit frustrating in this way. I'd love to be able to call `form.reset()` or use the `reset` button input and not have to worry about if I have controlled inputs in my form that also need to have their state updated.

However this PR is about fixing number inputs, not changing the controlled input specification. I can write up my thoughts on this in a separate issue.

nhunzaker commented:
@yaycmyk I filed an issue requesting support for `form.reset()` with controlled inputs:

https://github.com/facebook/react/issues/8011

Maybe it could get through, and that would eliminate one problem. Next would be tracking down what browser extensions have caused known issues with controlled inputs.

nhunzaker commented:
Okay! One thing down... For number inputs, only assigning the attribute when `input.validate.badInput` is `false` addresses issues where bad input causes Chrome to clear the number input.

Decimal places are the final thing... back to the salt mines.

nhunzaker commented:
Related (same issue). I did encounter a fun new one:

![plus-two](https://cloud.githubusercontent.com/assets/590904/19545019/1e6c9786-9651-11e6-8765-4b958867f835.gif)

In case it shows up on the issues.

It would be convenient if there was a way for me to see if the character length of the input is different than the reported value.

nhunzaker commented:
Hey.... I think I got something! 🎉 

What if we only updated number inputs if they both:

1. Do not contain bad input?
2. Are not in focus?

I've pushed that to this branch. It's working quite well:

![works](https://cloud.githubusercontent.com/assets/590904/19545601/9ae0dbfc-9655-11e6-8fcd-2994e148dfc3.gif)

As far as I can tell, there's simply no way to get around the attribute assignment issues in Chrome. It's just going to drop trailing decimal places... 

But! We can at least avoid that while the user is typing into the input. Of course, there are caveats:

1. Number inputs will only update their attribute when they don't have focus
2. This attaches an `onBlur` "wrapper" handler, similar to `onChange`

But pretty good, all things considered. What do you think?

zdzinks commented:
---------- Wiadomość przekazana dalej ----------
Od: *Trey Shugart* <notifications@github.com>
Data: czwartek, 20 października 2016
Temat: [facebook/react] Bypass synthetic event system for Web Component
events (#7901)
Do: facebook/react <react@noreply.github.com>


I'd like it better if we could just pass through all props to element
properties but it seems like that ship has sailed since most web components
aren't designed to handle properties properly. A massive loss to the
community IMO.

It feels like that statement is clumping all web components into a single
bag of poor design, which in many cases they are, but it doesn't mean you
can't optimise for the ones that are designed well. A couple paradigms I'm
trying to push in the community (and with SkateJS
<https://github.com/skatejs/skatejs>) are:

   - events up, props down
   - use props as your source of truth, sync attributes to props when you
   can

Rob Dodson placed emphasis on the latter in his Polymer Summit talk, so I
think that's something they're trying to espouse, too. It's unfortunate the
primitives don't make this more obvious, but I think that comes with the
nature of most browser built-ins these days.

React not setting props, and not supporting custom events, is the reason
we've had to maintain that React integration library I posted above (
https://github.com/webcomponents/react-integration). It's worth looking at
as a source of some patterns that are definitely working for us in
production. It's also worth noting that the patterns employed there are
also used in Skate's wrapper around Incremental DOM
<https://github.com/google/incremental-dom>. We set props for everything we
can, falling back to attributes as a last resort.

In that lib, events have their own special convention, similar to React's.
This in particular is something to pay attention to because adding event
listeners if prop is a function - without checking the name for something
like /^on[A-Z].*/ - will prefer events over properties.You could flip that
around and set props if prop in element and leave it up to the web
component to addEventListener() when the onClick property is set. This
creates a little more work for the component, but means you don't have to
do the prefix checking if you're not a fan.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/facebook/react/issues/7901#issuecomment-255018046>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AVj1bZeRAUwiMKhn2FuPqtsaY9e9ArnGks5q1wbkgaJpZM4KQUs->
.



-- 
null

nhunzaker commented:
@zdzinks did you mean to post this here? https://github.com/facebook/react/issues/7901

nhunzaker commented:
Gentle prod. 

Is this approach tenable? An input's `value` attribute will not be set on bad input or if the input is focused. On blur, it will assign the attribute (but only if it isn't bad input).

gaearon commented:
This looks reasonable to me but I have no context over all the back and forth about these decisions on inputs. I'll assign this to myself so it doesn't get lost but it will take me a while to be able to get back to reviewing this. It would help if you provided the behavior table before/after this change for major browser.

nhunzaker commented:
Thanks :). There's probably a table in here some where... I'll dig it up.

gaearon commented:
That table would be outdated with new approach wouldn't it?

nhunzaker commented:
@gaearon Ah I messed up. I meant to say that it wouldn't be a big deal.

nhunzaker commented:
@gaearon Okay! 

facebook:master
http://natehunzaker.com/react-issue-reproducer/cases/input/
https://github.com/nhunzaker/react-issue-reproducer/blob/gh-pages/cases/input/index.html

nhunzaker:nh-chrome-backspace:
http://natehunzaker.com/react-issue-reproducer/edge/cases/input/
https://github.com/nhunzaker/react-issue-reproducer/blob/gh-pages/edge/cases/input/index.html

## Is decimal place lost when backspacing a number input from "3.1" to "3."?

![test](https://cloud.githubusercontent.com/assets/590904/19747862/3cc5cf6c-9bac-11e6-91f3-7be8002e3e89.gif)

| Browser        | facebook:master  | nh-chrome-backspace           |
| -------------- | ---------------- | ----------------------------- |
| Chrome 54*      | lost             | maintained                    |
| Firefox 48     | maintained       | maintained                    |
| Safari 10      | lost             | maintained                    |
| IE 11          | lost             | maintained                    |
| MS Edge 14     | lost             | maintained                    |

The decimal place is dropped in Chrome when "3." the input loses focus. There's not much we can do here, and I don't think there is anything we can do about it unless we move away from always setting the `value` attribute.

## Does pushing "e" at the end of a number like "3 . 1 4 e", cause the input value to be lost?

![test](https://cloud.githubusercontent.com/assets/590904/19747869/4bfa3766-9bac-11e6-9720-ea57a353d0ac.gif)

| Browser        | facebook:master  | nh-chrome-backspace           |
| -------------- | ---------------- | ----------------------------- |
| Chrome 54      | lost             | maintained                    |
| Firefox 48     | maintained       | maintained                    |
| Safari 10      | lost             | maintained                    |
| IE 11*         | impossible       | impossible                    |
| MS Edge 14     | maintained       | maintained                    |

* IE11 does not seem to allow bad input typed into the end of a number

## Does pushing "ee" in the middle of 3.14 (like 3.ee14) preserve the display value?

![test](https://cloud.githubusercontent.com/assets/590904/19747876/5c61d78a-9bac-11e6-8aee-dc2491c59431.gif)

| Browser        | facebook:master                                 | nh-chrome-backspace            |
| -------------- | ----------------                                | -----------------------------  |
| Chrome 54      | no (empty) | yes                            |
| Firefox 48     | no: 300000000000000e                            | yes                            |
| Safari 10      | no: (first e: 300000000000000, second e: empty) | yes                            |
| IE 11          | impossible                                      | first e: yes, second e: no (3) |
| MS Edge 14     | no: (first e: 300000000000000, second e: empty) | yes                            |

---

Is this roughly what you were thinking?

nhunzaker commented:
Something interesting is that, since starting this feature, I've upgraded to macOS Sierra. Safari 10 consistently produces the same exact edge cases as Chrome.

gaearon commented:
What about this case:

* Delete . from 3.00

I think I saw issues with it before.

nhunzaker commented:
@gaearon While very difficult to get "3.00" to display, I did not see any issues with removing the decimal place. I _did_ see issues adding a decimal place and typing zeros after inserting a decimal place. Both are fixed in this PR and I've added tables for them in my report (and below):

## Does typing "3 . 0" preserve the trailing ". 0"?

![test](https://cloud.githubusercontent.com/assets/590904/19829258/eebe277e-9da9-11e6-9297-e7431a001825.gif)

| Browser        | facebook:master                        | nh-chrome-backspace           |
| -------------- | ----------------                       | ----------------------------- |
| Chrome 54      | no                                     | yes                           |
| Firefox 48     | no                                     | yes                           |
| Safari 10      | no (wont even allow the decimal point) | yes                           |
| IE 11          | no (wont even allow the decimal point) | yes                           |
| MS Edge 14     | no (wont even allow the decimal point) | yes                           |
|                |                                        |                               |

## Does inserting a decimal place to make "300" into "3.00" maintain the trailing zeroes?

![test](https://cloud.githubusercontent.com/assets/590904/19829325/3f1b162c-9dab-11e6-9a4c-e86c3b2e1a3e.gif)

| Browser        | facebook:master  | nh-chrome-backspace           |
| -------------- | ---------------- | ----------------------------- |
| Chrome 54      | no               | yes                           |
| Firefox 48     | no               | yes                           |
| Safari 10      | no               | yes                           |
| IE 11          | no               | yes                           |
| MS Edge 14     | no               | yes                           |

nhunzaker commented:
Conflicts resolved with recent `onChange` updates to `ReactDOMInput`

joh1234 commented:
Update README
https://goo.gl/JBljdR

qagiux commented:
Gonna do the first half of this in  https://goo.gl/JBljdR

nhunzaker commented:
@gaearon: Updated with master and re-verified using:

React Stable (facebook/react):
http://natehunzaker.com/react-issue-reproducer/cases/input/

Fix (nhunzaker/react#nh-chrome-backspace):
http://natehunzaker.com/react-issue-reproducer/edge/cases/input/

I understand that this probably takes a significant time commitment to validate, but I really want to get this fixed, so I also went ahead and made a case with many other inputs:

React Stable (facebook/react):
http://natehunzaker.com/react-issue-reproducer/cases/input-kitchen-sink/

Fix (nhunzaker/react#nh-chrome-backspace):
http://natehunzaker.com/react-issue-reproducer/edge/cases/input-kitchen-sink/

Please let me know if there is anything else I can do to speed this along.

ruiaraujo commented:
@nhunzaker Thanks for this, this bug is very annoying.
@gaearon Please save us from from bad number inputs. 😄

gaearon commented:
Thank you very much. I’m sorry I’m not getting enough time to review this. We are very focused on Fiber right now, but we’ll get around to the long trail of issues once it’s more stable.

nhunzaker commented:
Understood. I appreciate the response, that's very helpful to know. I guess I need to figure out how to help out on Fiber!

gaearon commented:
https://github.com/facebook/react/issues/7925

Aweary commented:
I'd be happy to review this again, at the very least to help your review go quicker @gaearon

gaearon commented:
@Aweary If you could review and help get this in, it would be awesome.

Aweary commented:
@nhunzaker can you point me to where you expect `inst._wrapperState.onBlur` to be called? From what I can tell we're expecting the `onBlur` function to be called when the input is blurred, but using your kitchen sink example (thanks for those btw!) I'm not seeing the `value` attribute update at all for number inputs and `handleBlur` is not being called.

nhunzaker commented:
![nooo](https://media.giphy.com/media/26hirThH4g9KmZ5gk/source.gif) I see that too. Sheesh, I'll reconfirm this. It's possibly a regression with a recent rebase.

Sorry... I'm looking into this now.

Aweary commented on src/renderers/dom/stack/client/wrappers/ReactDOMInput.js:
```diff
@@ -122,6 +123,7 @@ var ReactDOMInput = {
       initialChecked: props.checked != null ? props.checked : props.defaultChecked,
       initialValue: props.value != null ? props.value : defaultValue,
       listeners: null,
+      onBlur: _handleBlur.bind(inst),
```
We need to make sure this is invoked when the input is blurred, and add a test if possible.

nhunzaker replied:
This won't get called anymore. Though I'm unclear why. I had thought this function was a filter/transform step for props before they entered the component. When I follow the stack trace for the blur event, this doesn't seem to be the case.

spicyj replied:
It sort of is but we read event handlers from the original props now. All the custom logic is in ChangeEventPlugin and restoreControlledState.

Aweary commented:
@nhunzaker thanks, I'll look into it as well. Added a review with changes requested in the meantime.

nhunzaker commented:
Awesome. Ya, I need to dig into the change that happened when the `onChange` wrapper was removed. The blur handler is no longer getting executed.

nhunzaker commented:
I just confirmed that my `onBlur` handler approach is no longer valid with the new way event handlers are called in React (which is a super cool change, just unfortunate for me).

I've pushed a few tests. The build fails, as it should. I like the technique of listening to onBlur, in principle. Now to just figure out how to apply it.

nhunzaker commented:
Okay...

I moved the onBlur rule for number inputs to the ChangeEventPlugin. The value attribute updates correctly, as seen in the controlled input here:

http://natehunzaker.com/react-issue-reproducer/edge/cases/input/index.html

Also I upstreamed with master and re-recorded the fiber tests.

I don't feel great about _where_ this onBlur rule lives, but I'll point that out in the file diff in a moment.

nhunzaker commented on src/renderers/dom/shared/eventPlugins/ChangeEventPlugin.js:
```diff
@@ -351,6 +351,11 @@ var ChangeEventPlugin = {
+    if (topLevelType === 'topBlur' && targetNode.type === 'number') {
+      targetNode.setAttribute('value', targetNode.value);
+    }
   },
```
Basically, onBlur just sets the target node's value attribute to it's current value, but only if it's a number input. I don't know that this is the best place to put this. Any suggestions? 

cc @spicyj ?

spicyj replied:
I'm not sure if there is a better place, sorry.

nhunzaker commented on src/renderers/dom/shared/wrappers/__tests__/ReactDOMInput-test.js:
```diff
@@ -1028,4 +1039,59 @@ describe('ReactDOMInput', () => {
+      expect(node.getAttribute('value')).toBe('2');
+    });
+  });
+
```
Does this sufficiently cover this feature? 

1. Coverage of non-text inputs
2. Coverage of focused number inputs
3. Coverage of focused number inputs that get blurred

Any edge cases I'm not thinking of?

nhunzaker replied:
One. Uncontrolled inputs... Writing a test case for that now.

nhunzaker replied:
Okay. Covered. Though I had to expose `controlled` as a wrapper state value.

nhunzaker commented:
Continuing to do testing, but I've updated this code to properly accomodate uncontrolled inputs. I did this by moving the `controlled` wrapper state assignment to be a general flag. Whenever the ChangeEventPlugin detects a blur on a controlled number input, it will assign the value attribute to that of the current input value.

This gets everything to pass on both the standard DOM renderer and React Fiber. 

Test case:
http://natehunzaker.com/react-issue-reproducer/edge/cases/input/index.html

jquense commented:
next time folks ask me why I wrote a number picker widget in react-widgets instead of using the native one, i'm going to point them to this issue :P

nhunzaker commented:
@jquense it has been quite a journey :P

aweary commented on src/renderers/dom/shared/eventPlugins/ChangeEventPlugin.js:
```diff
@@ -289,6 +289,18 @@ function getTargetInstForInputOrChangeEvent(
+  }
+
+  // If controlled, assign the value attribute to the current value on blur
+  node.setAttribute('value', '' + node.value);
```
Won't this cause us to set the `value` attribute every time the input is blurred, even if the value hasn't changed?

nhunzaker replied:
Hmm. Yes. I'll through a conditional around this that does a comparison of the value property. 

Could the input value tracker help us out here to avoid DOM access? cc @jquense

aweary commented on src/renderers/dom/fiber/wrappers/ReactDOMFiberInput.js:
```diff
@@ -132,11 +132,8 @@ var ReactDOMInput = {
-
-    if (__DEV__) {
-      node._wrapperState.controlled = isControlled(props);
-    }
```
Hmm, I wonder why this was dev-only in the first place?

sebmarkbage replied:
It was only used for warnings. We intentionally want to minimize state and memory usage associated with controlled components. The goal is to get rid of the entire wrapper object making this a prod dependency makes that harder.

nhunzaker commented:
@aweary 

- Fixed lint Issues.
- Rebased with Master. 
- Added a conditional around the value attribute assignment. 

I've also updated the test case:
http://natehunzaker.com/react-issue-reproducer/edge/cases/input/index.html

aweary commented:
@nhunzaker I'm reviewing this today and ran into an issue that might be difficult to address. With this change, it makes it impossible to enter values smaller than `0.1` because [`0.0` coerces to `0`](https://github.com/facebook/react/pull/7359/files#diff-f47152d575264e77822c1f3abc1a6342R177), so you cannot type numbers like `0.01`.

gaearon commented:
(Good case for a fixture)

aweary commented on src/renderers/dom/stack/client/wrappers/ReactDOMInput.js:
```diff
@@ -174,14 +171,17 @@ var ReactDOMInput = {
-      // To avoid side effects (such as losing text selection), only set value if changed
-      if (newValue !== node.value) {
-        node.value = newValue;
+      if (value === 0) {
```
@nhunzaker can we check `node.value` here too and see if its `"0.0"`? Since `0.0 === 0` I think we're going to need to rely on checking the stringified value of the node's value.

nhunzaker replied:
Huh... was there a recent fix for inputs in Chrome? I can assign `node.value = "0.0000"` and it totally just works... I don't believe that was ever the case before.

nhunzaker commented:
Man... `0.0`... rough. Alright. Here goes...

aweary commented:
> Unable to enter floating point numbers less than 0.0 as 0.0 is being coerced to 0

This can be addressed as mentioned in my other comment (https://github.com/facebook/react/pull/7359#discussion_r95198269), I tested the change locally in Chrome (OS X) and it seemed to fix it.

> This may be a separate issue with the event system, but handleControlledInputBlur throws in IE9 as targetInst is null or undefined and it attempts to access targetInst._wrapperState

I don't think `targetInst` is ever supposed to be `null` / `undefined`, so I'm not sure what's going on here. We may need to do more extensive testing here. If `targetInst` is allowed to be `null`/undefined` then we'll need to adjust our approach.

nhunzaker commented:
@aweary Updated, though I can't reproduce the `0.001` precision issue. What browsers were you encountering it with?

nhunzaker commented:
Additionally, I've rebased with master and added a browser test fixture for number inputs.

nhunzaker commented:
@aweary sorry for all the noise. Looks like I wasn't running `parseFloat` in my number input's `onChange` callback... I can see the case now.

jquense commented:
:thumbsup: if you don't mind I'll use em as a template for extracting out some fixture components

nhunzaker commented:
@jquense sounds good to me.

nhunzaker commented:
Whew! @aweary really great catch. I've made a change (0c48709) that addresses the precision issue (typing anything more than `0.0`). I've also added test coverage for that case, and an extra test for transitioning from 0 to a blank value. 

I'll start investigating:

- This may be a separate issue with the event system, but handleControlledInputBlur throws in IE9 as targetInst is null or undefined and it attempts to access targetInst._wrapperState

- Cannot enter numbers in exponential format (2e4) in IE10/IE11

nhunzaker commented:
> Cannot enter numbers in exponential format (2e4) in IE10/IE11

Fixed this with https://github.com/facebook/react/pull/7359/commits/e65151bacd53d6d5df084f58a34faaa835d127e7

This fix also had the side-effect of allowing bad input in IE10+, which I had previously believed to be a browser quirk. You can now type bad input all you want in IE without a fuss

I've tested this fix against IE9-14, OSX Chrome, OSX Firefox, and Safari 10

Final issue to work through is:

> This may be a separate issue with the event system, but handleControlledInputBlur throws in IE9 as targetInst is null or undefined and it attempts to access targetInst._wrapperState

nhunzaker commented:
I've fixed the IE9/10 exception with c34bf15. Worth noting that the `value` attribute doesn't appear to update unless you refresh the console:

![ie](https://cloud.githubusercontent.com/assets/590904/22176289/dce62660-dfd4-11e6-84ca-18cc0aa86b03.gif)

It through me for a loop anyway.

nhunzaker commented:
So that's all done. Looks like Fiber's failing my input 0 to blank string transition tests. Time to sort that out...

nhunzaker commented:
Ah, got it. I didn't realize `ReactDOMFiberInput.js` was a near copy of `ReactDOMInput.js`. I just had to copy over the new logic. I've done this in 55f7e57. 

@aweary I believe this is ready for another review.

nhunzaker commented on src/renderers/dom/shared/eventPlugins/ChangeEventPlugin.js:
```diff
@@ -289,6 +289,26 @@ function getTargetInstForInputOrChangeEvent(
 }
 
+function handleControlledInputBlur(inst, node) {
+  // TODO: In IE, inst is occasionally null. Why?
```
I threw a TODO in here. This gets called later down the road. I'm unclear why, in IE9/10 only, it's getting called with a nully instance.

nhunzaker commented:
@aweary Conflicts resolved and fiber tests re-recorded. 

I actually think this probably doesn't depend on #8575 because this targets master. Is there anything left here?

nhunzaker commented:
Actually, I just squashed down the lint commits. _Now_ I think it's ready. :)

nhunzaker commented:
Well... disregard my comment about #8575. _One_ of these needs to get merged first. Is there anything I can do to help move things forward?

aweary commented:
@nhunzaker can you resolve the conflicts with the dom fixtures? I'll try to review this today! 

> Well... disregard my comment about #8575. One of these needs to get merged first. Is there anything I can do to help move things forward?

This shouldn't rely on #8575 because its not targeting master as you said, right? Or am I missing something?

nhunzaker commented:
@aweary Sorry for the delay... I went kind of heads down today. I've resolved the conflicts and updated the number input test fixtures to use the components we made in the fixture work you just merged.

iangilman commented:
I found my way here after isolating an issue with the number input... I'm hoping this patch fixes the issue I found, but I'm writing here to confirm it's been covered.

In Chrome if you have a React 15.4.2 number input with a value and you select the value and then type `-` the value is removed but no `-` is added. Here's a simple repro: http://codepen.io/iangilman/pen/EWWgYp

If this issue is not covered in this patch I'm happy to file a new issue, but I thought I should check here first!

nhunzaker commented:
Ooo nice. I'll add it to the text cases

nhunzaker commented:
Rebased with master. I also added @iangilman's case for trailing dashes, and a test case for negative numbers:

![dashes](https://cloud.githubusercontent.com/assets/590904/24322979/3c53a53c-1144-11e7-95e2-b44f868d411b.gif)

jquense commented:
I'm interested in why `3-` needs to be supported (if only to make sure react-widgets does it)

nhunzaker commented:
@jquense You know what, I think I have this wrong. The test case is when you select the entire number value and `-` to start typing a new number. Stable react sees `-` as an invalid number and wipes away the value.

I'll update the test case. I also need to figure out what to do about `prettier` formatting. `prettier` wants to put my ESLint ignore comments on a new line, which causes linting to fail.

Away I go.

jquense commented:
ah gotcha. If it helps btw please feel free to look at https://github.com/jquense/react-widgets/blob/v4/packages/react-widgets/src/NumberInput.js to see how i fumbled through on this logic

nhunzaker commented:
Cool. I just had to update the test case, everything works fine. If the number isn't valid, it parses to a blank value, which is also what the number input reports, so no change occurs.

I've also run prettier on this branch and dealt with the associated eslint issue.

nhunzaker commented:
@jquense This NumberInput is cool. How crazy would it be to bake something like this into React?

Later... later.. I want to get this PR merged so badly, haha.

sebmarkbage commented on src/renderers/dom/shared/eventPlugins/ChangeEventPlugin.js:
```diff
@@ -258,6 +258,26 @@ function getTargetInstForInputOrChangeEvent(topLevelType, targetInst) {
+  // If controlled, assign the value attribute to the current value on blur
+  let value = '' + node.value;
+  if (node.getAttribute('value') !== value) {
+    node.setAttribute('value', value);
```
Sorry, I'm not caught up on the entire history here.

Why is this special case needed? extractEvents should not have side-effects in it.

Can this be handled by the normal controlledness flow in updateWrapper?

sebmarkbage replied:
Would it be enough to flag this with `enqueueStateRestore` so that the normal `updateWrapper` kicks in at the end of the blur event?

nhunzaker replied:
I can verify. As for history:

React controlled inputs sync up the value attribute with the value property. This prevents state mismatches with form.reset() and browser plugins (though i never got a reply on which ones). Unfittingly, on number inputs in Chrome and Safari, this forces a check on the input, removing trailing decimal places, invalid characters, etc. 

The only way i could get around this (i tried for months, off and on) was to defer setting the value attribute on blur.

I haven't done much with enqueueStateRestore. I'll give it a go.

sebmarkbage replied:
IMO we should just not set the `value` attribute. People writing manual DOM operations don't do that and it works fine for them, so not sure why we do.

Would just avoiding setting the attribute all together fix the problem?

nhunzaker replied:
That would be lovely. The only edge case is form.reset(). Honestly, I'd be in favor of just adding a warning about this, or enqueing updateWrapper on the reset form event if it can be done.

sebmarkbage replied:
We already set defaultValue which helps `form.reset()`, right?

I agree that form reset could enqueue updateWrapper. It's kind of a weird use case with any controlled component though.

nhunzaker replied:
Assigning defaultValue has the effect of updating the value attribute. Which puts us in the same boat. the only trick with supporting form reset would be the execution order. Last I checked, reset emits _before_ the input actually changes. We'd have to do some sort of next tick behavior. I'm curious what the order of execution would look like to a user in the browser. Would it be transparent?

Something I can figure out.

nhunzaker replied:
I have an issue out for this, but we never finished the discussion:
https://github.com/facebook/react/issues/8011

sebmarkbage replied:
Honestly, I'd rather just disable form reset support in React (either hard failure or just a warning) until we have a simpler fix specific to resets. What do you think about that?

aweary replied:
There were apparently some other reasons for setting the `value` attribute that weren't fully explained, see https://github.com/facebook/react/pull/7359#issuecomment-244768124. Maybe @jimfb or @spicyj have some historical insight that would be useful?

aweary replied:
We also need to be clear that `form.reset` is totally fine when the form contains only uncontrolled inputs. If we did go the route of warning it should only warn if the form contains one or more controlled inputs.

spicyj replied:
I think just .reset() and CSS `input[value=foo]` selectors. TBH the latter was a pretty common feature request so I'd be sad to see it go.

sebmarkbage replied:
There are other attributes that matter. Such as all of those that don't have corresponding properties (generally things outside the HTML spec). There are edge cases that need removing attributes (like `progress`). However, for `value` (or `checked`) I don't believe there were any cases that needed it outside of intuition. E.g. the attribute shows up in DOM dev tools in Chrome etc.

However, there is nothing that relies on it being there because all those thousands of websites that do manual DOM management. Nobody manually calls `setAttribute` in those cases. However, as we've seen, there are cases where setting it all the time is unusual such as in this case. So it leads to undiscovered browser deviations.

That's why I think our design should be to model what a user would write manually - even if that seems unintuitive of how people think React works.

nhunzaker replied:
In a most cases I can think of with manual manipulation it's actually undesirable to set the value attribute because you want to rely on form.reset. at least, this has always been the case for me. Being able to style text inputs based on a dynamic value attribute is a privilege react provides, one I would be comfortable with no longer having. 

React doing Dom manipulation, the correct way, is the abstraction I've always had. I think this is a very strong metaphor that makes it easier to reason about what react is doing. We deviate from that with controlled inputs. I'm in favor of this change, which moves closer towards that idea.

So I agree, armchair theory aside 😂. I can start scaffolding out what that would look like, but I won't be able to pick it up until Monday, if someone else wants to give it a shot.

nhunzaker replied:
Oof. I finally did it. Please see:

https://github.com/facebook/react/pull/10150

btoueg commented:
I've bumped my project to 15.5.4 and the issue is still there.

gaearon commented:
@btoueg Provide a reproducing case please?

btoueg commented:
Sorry for the delay. My report was a false negative. Works fine now, congrats on fixing the issue.

hyperknot commented:
I've run into the following issue, I believe it is related to this.

The following code example works well with `text` input type, but allows inputs like `012` or `0012` to be entered when using the `number` input type.

The `console.log` line always runs and shows the right value, and the App state is also correct when checked with React Developer Tools. It's only that the controlled input is not being "controlled" somehow.

Why is this happening, and what is the recommended way to use number inputs with React? 

    class App extends Component {
      constructor() {
        super()
        this.state = {
          value: '',
        }
      }
    
      handleChange = e => {
        const value = e.target.value
        const num = parseInt(value, 10)
        console.log(num)
        this.setState({ value: isNaN(num) ? '' : num })
      }
    
      render() {
        return (
          <div className="App">
            <input type="number" value={this.state.value} onChange={this.handleChange} />
          </div>
        )
      }
    }

Browser: Chrome Version 60.0.3112.101, OS X 10.12.6
